File: typescript/src/builders.ts
/**
 * Builders for the attestation, PoP and DPoP JWTs (draft-ietf-oauth-attestation-based-client-auth).
 */
import { randomUUID } from "node:crypto";
import type { JWK } from "jose";

import { SigningKeyPair, requireText, signCompact } from "./signing-key-pair.js";
import { type JwsSigner, signExternal } from "./signer.js";

export const ATTESTATION_TYP = "oauth-client-attestation+jwt";
export const POP_TYP = "oauth-client-attestation-pop+jwt";
export const DPOP_TYP = "dpop+jwt";

function nowSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

/**
 * Builds a Client Attestation JWT — the credential a Client Attester issues to name a client
 * (`sub`) and bind its instance key via `cnf.jwk`. Attester side; sign with the attester's key.
 */
export class ClientAttestationBuilder {
  readonly #attesterKey: SigningKeyPair | JwsSigner;
  readonly #issuer: string;
  #clientId: string | null = null;
  #cnfJwk: JWK | null = null;
  #issuedAt: number | null = null;
  #expiresAt: number | null = null;
  #ttl: number | null = null;
  #authorizationDetails: unknown[] | null = null;
  #workload: Record<string, unknown> | null = null;

  constructor(attesterKey: SigningKeyPair | JwsSigner, issuer: string) {
    this.#attesterKey = attesterKey;
    this.#issuer = requireText(issuer, "issuer");
  }

  /** The client being attested — becomes the attestation `sub` (= `client_id`). */
  clientId(clientId: string): this {
    this.#clientId = clientId;
    return this;
  }

  /** Binds the client instance key as `cnf.jwk`. Must be a public JWK. */
  confirmationJwk(publicInstanceJwk: JWK): this {
    this.#cnfJwk = publicInstanceJwk;
    return this;
  }

  /** Convenience: binds the public half of the given instance key as `cnf.jwk`. */
  confirmationKey(instanceKey: SigningKeyPair): this {
    return this.confirmationJwk(instanceKey.publicJwk());
  }

  issuedAt(epochSeconds: number): this {
    this.#issuedAt = epochSeconds;
    return this;
  }

  /** Sets an absolute `exp`. */
  expiresAt(epochSeconds: number): this {
    this.#expiresAt = epochSeconds;
    return this;
  }

  /** Sets `exp` to `iat + seconds`. */
  expiresIn(seconds: number): this {
    this.#ttl = seconds;
    return this;
  }

  /** The optional RFC 9396 `authorization_details` entitlement the attester asserts. */
  authorizationDetails(details: unknown[]): this {
    this.#authorizationDetails = details;
    return this;
  }

  /** Optional attester-asserted `workload` attributes. */
  workload(workload: Record<string, unknown>): this {
    this.#workload = workload;
    return this;
  }

  async build(): Promise<string> {
    const sub = requireText(this.#clientId, "client_id");
    if (this.#cnfJwk === null) {
      throw new Error("confirmation key (cnf.jwk) is required");
    }
    const iat = this.#issuedAt !== null ? this.#issuedAt : nowSeconds();
    const exp = this.#resolveExpiry(iat);
    const claims: Record<string, unknown> = {
      iss: this.#issuer,
      sub,
      iat,
      exp,
      cnf: { jwk: this.#cnfJwk },
    };
    if (this.#authorizationDetails && this.#authorizationDetails.length > 0) {
      claims.authorization_details = this.#authorizationDetails;
    }
    if (this.#workload && Object.keys(this.#workload).length > 0) {
      claims.workload = this.#workload;
    }
    if (this.#attesterKey instanceof SigningKeyPair) {
      return signCompact(claims, this.#attesterKey, ATTESTATION_TYP, false);
    }
    return signExternal(claims, this.#attesterKey, ATTESTATION_TYP);
  }

  #resolveExpiry(iat: number): number {
    if (this.#expiresAt !== null) {
      return this.#expiresAt;
    }
    if (this.#ttl !== null) {
      return iat + this.#ttl;
    }
    throw new Error("expiry is required: call expiresAt(...) or expiresIn(...)");
  }
}

/**
 * Builds a Client Attestation PoP JWT proving possession of the instance key. Client side of
 * `attest_jwt_client_auth`; mint a fresh one per token request.
 */
export class PopBuilder {
  readonly #instanceKey: SigningKeyPair;
  #clientId: string | null = null;
  #audience: string | null = null;
  #challenge: string | null = null;
  #jwtId: string | null = null;
  #issuedAt: number | null = null;

  constructor(instanceKey: SigningKeyPair) {
    this.#instanceKey = instanceKey;
  }

  /** Sets the optional `iss`; when present the verifier requires it to equal the attestation `sub`. */
  clientId(clientId: string | null): this {
    this.#clientId = clientId;
    return this;
  }

  /** The AS identifier this PoP is bound to (`aud`) — required. */
  audience(audience: string): this {
    this.#audience = audience;
    return this;
  }

  challenge(challenge: string | null): this {
    this.#challenge = challenge;
    return this;
  }

  jwtId(jwtId: string): this {
    this.#jwtId = jwtId;
    return this;
  }

  issuedAt(epochSeconds: number): this {
    this.#issuedAt = epochSeconds;
    return this;
  }

  async build(): Promise<string> {
    if (!this.#audience) {
      throw new Error("audience (aud) is required");
    }
    const claims: Record<string, unknown> = {
      aud: this.#audience,
      jti: this.#jwtId ?? randomUUID(),
      iat: this.#issuedAt !== null ? this.#issuedAt : nowSeconds(),
    };
    if (this.#clientId) {
      claims.iss = this.#clientId;
    }
    if (this.#challenge) {
      claims.challenge = this.#challenge;
    }
    return signCompact(claims, this.#instanceKey, POP_TYP, false);
  }
}

/**
 * Builds a DPoP proof JWT (RFC 9449) for attestation combined mode
 * (`attest_jwt_client_auth_dpop`): the embedded `jwk` header MUST be the attestation's `cnf` key.
 */
export class DpopProofBuilder {
  readonly #instanceKey: SigningKeyPair;
  #htm = "POST";
  #htu: string | null = null;
  #nonce: string | null = null;
  #jwtId: string | null = null;
  #issuedAt: number | null = null;

  constructor(instanceKey: SigningKeyPair) {
    this.#instanceKey = instanceKey;
  }

  /** The token request's HTTP method (`htm`); defaults to `POST`. */
  method(htm: string | null | undefined): this {
    if (htm) {
      this.#htm = htm;
    }
    return this;
  }

  /** The token endpoint URL (`htu`) — required. */
  uri(htu: string): this {
    this.#htu = htu;
    return this;
  }

  /** The server challenge, carried in DPoP `nonce`. */
  nonce(nonce: string | null): this {
    this.#nonce = nonce;
    return this;
  }

  jwtId(jwtId: string): this {
    this.#jwtId = jwtId;
    return this;
  }

  issuedAt(epochSeconds: number): this {
    this.#issuedAt = epochSeconds;
    return this;
  }

  async build(): Promise<string> {
    if (!this.#htu) {
      throw new Error("uri (htu) is required");
    }
    const claims: Record<string, unknown> = {
      htm: this.#htm,
      htu: this.#htu,
      jti: this.#jwtId ?? randomUUID(),
      iat: this.#issuedAt !== null ? this.#issuedAt : nowSeconds(),
    };
    if (this.#nonce) {
      claims.nonce = this.#nonce;
    }
    return signCompact(claims, this.#instanceKey, DPOP_TYP, true);
  }
}

File: go/builders.go
package clientattestation

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// JWT "typ" header values for the three artifact kinds.
const (
	AttestationTyp = "oauth-client-attestation+jwt"
	PopTyp         = "oauth-client-attestation-pop+jwt"
	DpopTyp        = "dpop+jwt"
)

// ClientAttestationBuilder builds a Client Attestation JWT — the credential a
// Client Attester issues to name a client (sub = client_id) and bind its
// instance key via the RFC 7800 cnf.jwk claim. This is the attester side: sign
// with the attester's issuing key.
type ClientAttestationBuilder struct {
	attesterKey          *SigningKeyPair
	attesterSigner       JwsSigner
	issuer               string
	clientID             string
	cnfJWK               map[string]any
	issuedAt             *int64
	expiresAt            *int64
	ttl                  *int64
	authorizationDetails []map[string]any
	workload             map[string]any
}

// NewClientAttestationBuilder starts a builder signing with attesterKey and
// asserting the given issuer (attestation "iss").
func NewClientAttestationBuilder(attesterKey *SigningKeyPair, issuer string) *ClientAttestationBuilder {
	return &ClientAttestationBuilder{attesterKey: attesterKey, issuer: issuer}
}

// NewClientAttestationBuilderWithSigner starts a builder signing the attestation with an external
// JwsSigner (e.g. an *OpenBaoTransitSigner whose key lives in a vault) instead of a local SigningKeyPair.
func NewClientAttestationBuilderWithSigner(signer JwsSigner, issuer string) *ClientAttestationBuilder {
	return &ClientAttestationBuilder{attesterSigner: signer, issuer: issuer}
}

// ClientID sets the client being attested — the attestation "sub" (= client_id).
func (b *ClientAttestationBuilder) ClientID(clientID string) *ClientAttestationBuilder {
	b.clientID = clientID
	return b
}

// ConfirmationJWK binds a public instance JWK as cnf.jwk.
func (b *ClientAttestationBuilder) ConfirmationJWK(publicInstanceJWK map[string]any) *ClientAttestationBuilder {
	b.cnfJWK = publicInstanceJWK
	return b
}

// ConfirmationKey binds the public half of the given instance key as cnf.jwk.
func (b *ClientAttestationBuilder) ConfirmationKey(instanceKey *SigningKeyPair) *ClientAttestationBuilder {
	return b.ConfirmationJWK(instanceKey.PublicJWK())
}

// IssuedAt sets an explicit "iat" (epoch seconds).
func (b *ClientAttestationBuilder) IssuedAt(epochSeconds int64) *ClientAttestationBuilder {
	b.issuedAt = &epochSeconds
	return b
}

// ExpiresAt sets an absolute "exp" (epoch seconds).
func (b *ClientAttestationBuilder) ExpiresAt(epochSeconds int64) *ClientAttestationBuilder {
	b.expiresAt = &epochSeconds
	return b
}

// ExpiresIn sets "exp" to iat + seconds.
func (b *ClientAttestationBuilder) ExpiresIn(seconds int64) *ClientAttestationBuilder {
	b.ttl = &seconds
	return b
}

// AuthorizationDetails sets the optional RFC 9396 authorization_details array.
func (b *ClientAttestationBuilder) AuthorizationDetails(details []map[string]any) *ClientAttestationBuilder {
	b.authorizationDetails = details
	return b
}

// Workload sets the optional attester-asserted workload object.
func (b *ClientAttestationBuilder) Workload(workload map[string]any) *ClientAttestationBuilder {
	b.workload = workload
	return b
}

// Build signs and returns the compact attestation JWT.
func (b *ClientAttestationBuilder) Build() (string, error) {
	issuer, err := requireText(b.issuer, "issuer")
	if err != nil {
		return "", err
	}
	sub, err := requireText(b.clientID, "client_id")
	if err != nil {
		return "", err
	}
	if b.cnfJWK == nil {
		return "", fmt.Errorf("confirmation key (cnf.jwk) is required")
	}
	iat := b.resolveIssuedAt()
	exp, err := b.resolveExpiry(iat)
	if err != nil {
		return "", err
	}
	claims := map[string]any{
		"iss": issuer,
		"sub": sub,
		"iat": iat,
		"exp": exp,
		"cnf": map[string]any{"jwk": b.cnfJWK},
	}
	if len(b.authorizationDetails) > 0 {
		claims["authorization_details"] = b.authorizationDetails
	}
	if len(b.workload) > 0 {
		claims["workload"] = b.workload
	}
	if b.attesterSigner != nil {
		return signExternal(claims, b.attesterSigner, AttestationTyp)
	}
	return b.attesterKey.signCompact(claims, AttestationTyp, false)
}

func (b *ClientAttestationBuilder) resolveIssuedAt() int64 {
	if b.issuedAt != nil {
		return *b.issuedAt
	}
	return time.Now().Unix()
}

func (b *ClientAttestationBuilder) resolveExpiry(iat int64) (int64, error) {
	if b.expiresAt != nil {
		return *b.expiresAt, nil
	}
	if b.ttl != nil {
		return iat + *b.ttl, nil
	}
	return 0, fmt.Errorf("expiry is required: call ExpiresAt(...) or ExpiresIn(...)")
}

// PopBuilder builds a Client Attestation PoP JWT proving possession of the
// instance key. Client side of attest_jwt_client_auth; mint a fresh one per
// token request and sign it with the instance key.
type PopBuilder struct {
	instanceKey *SigningKeyPair
	clientID    string
	audience    string
	challenge   string
	jwtID       string
	issuedAt    *int64
}

// NewPopBuilder starts a PoP builder signing with instanceKey.
func NewPopBuilder(instanceKey *SigningKeyPair) *PopBuilder {
	return &PopBuilder{instanceKey: instanceKey}
}

// ClientID sets the optional "iss" (= client_id).
func (b *PopBuilder) ClientID(clientID string) *PopBuilder {
	b.clientID = clientID
	return b
}

// Audience sets the AS identifier this PoP is bound to ("aud") — required.
func (b *PopBuilder) Audience(audience string) *PopBuilder {
	b.audience = audience
	return b
}

// Challenge sets the optional server challenge.
func (b *PopBuilder) Challenge(challenge string) *PopBuilder {
	b.challenge = challenge
	return b
}

// JwtID sets an explicit "jti" (otherwise a random UUID is used).
func (b *PopBuilder) JwtID(jwtID string) *PopBuilder {
	b.jwtID = jwtID
	return b
}

// IssuedAt sets an explicit "iat" (epoch seconds).
func (b *PopBuilder) IssuedAt(epochSeconds int64) *PopBuilder {
	b.issuedAt = &epochSeconds
	return b
}

// Build signs and returns the compact PoP JWT.
func (b *PopBuilder) Build() (string, error) {
	if _, err := requireText(b.audience, "audience (aud)"); err != nil {
		return "", err
	}
	iat := time.Now().Unix()
	if b.issuedAt != nil {
		iat = *b.issuedAt
	}
	jti := b.jwtID
	if jti == "" {
		jti = uuid.NewString()
	}
	claims := map[string]any{
		"aud": b.audience,
		"jti": jti,
		"iat": iat,
	}
	if b.clientID != "" {
		claims["iss"] = b.clientID
	}
	if b.challenge != "" {
		claims["challenge"] = b.challenge
	}
	return b.instanceKey.signCompact(claims, PopTyp, false)
}

// DpopProofBuilder builds a DPoP proof JWT (RFC 9449) for attestation combined
// mode (PoP method dpop_combined): the embedded "jwk" header MUST be the
// instance key bound in the attestation's cnf. Sign with the instance key.
type DpopProofBuilder struct {
	instanceKey *SigningKeyPair
	htm         string
	htu         string
	nonce       string
	jwtID       string
	issuedAt    *int64
}

// NewDpopProofBuilder starts a DPoP builder signing with instanceKey; htm
// defaults to "POST".
func NewDpopProofBuilder(instanceKey *SigningKeyPair) *DpopProofBuilder {
	return &DpopProofBuilder{instanceKey: instanceKey, htm: "POST"}
}

// Method sets the token request HTTP method ("htm"); blank is ignored.
func (b *DpopProofBuilder) Method(htm string) *DpopProofBuilder {
	if htm != "" {
		b.htm = htm
	}
	return b
}

// URI sets the token endpoint URL ("htu") — required.
func (b *DpopProofBuilder) URI(htu string) *DpopProofBuilder {
	b.htu = htu
	return b
}

// Nonce sets the optional server challenge, carried in DPoP "nonce".
func (b *DpopProofBuilder) Nonce(nonce string) *DpopProofBuilder {
	b.nonce = nonce
	return b
}

// JwtID sets an explicit "jti" (otherwise a random UUID is used).
func (b *DpopProofBuilder) JwtID(jwtID string) *DpopProofBuilder {
	b.jwtID = jwtID
	return b
}

// IssuedAt sets an explicit "iat" (epoch seconds).
func (b *DpopProofBuilder) IssuedAt(epochSeconds int64) *DpopProofBuilder {
	b.issuedAt = &epochSeconds
	return b
}

// Build signs and returns the compact DPoP proof JWT.
func (b *DpopProofBuilder) Build() (string, error) {
	if _, err := requireText(b.htu, "uri (htu)"); err != nil {
		return "", err
	}
	iat := time.Now().Unix()
	if b.issuedAt != nil {
		iat = *b.issuedAt
	}
	jti := b.jwtID
	if jti == "" {
		jti = uuid.NewString()
	}
	claims := map[string]any{
		"htm": b.htm,
		"htu": b.htu,
		"jti": jti,
		"iat": iat,
	}
	if b.nonce != "" {
		claims["nonce"] = b.nonce
	}
	return b.instanceKey.signCompact(claims, DpopTyp, true)
}
